pkg/cli/admin/release/extract.go

```diff
@@ -296,111 +299,64 @@ func (o *ExtractOptions) Run() error {
 		}
 	}
 
-	switch {
-	case len(o.File) > 0:
-		var manifestErrs []error
-		found := false
-		opts.TarEntryCallback = func(hdr *tar.Header, _ extract.LayerInfo, r io.Reader) (bool, error) {
-			if !o.ExtractManifests {
-				if hdr.Name != o.File {
-					return true, nil
-				}
-				if _, err := io.Copy(o.Out, r); err != nil {
-					return false, err
-				}
-				found = true
-				return false, nil
-			} else {
-				switch hdr.Name {
-				case o.File:
-					if _, err := io.Copy(o.Out, r); err != nil {
-						return false, err
-					}
-					found = true
-				case "image-references":
-					return true, nil
-				case "release-metadata":
-					return true, nil
-				default:
-					if ext := path.Ext(hdr.Name); len(ext) == 0 || !(ext == ".yaml" || ext == ".yml" || ext == ".json") {
-						return true, nil
-					}
-					klog.V(4).Infof("Found manifest %s", hdr.Name)
-					raw, err := ioutil.ReadAll(r)
-					if err != nil {
-						manifestErrs = append(manifestErrs, errors.Wrapf(err, "error reading file %s", hdr.Name))
-						return true, nil
-					}
-					ms, err := manifest.ParseManifests(bytes.NewReader(raw))
-					if err != nil {
-						manifestErrs = append(manifestErrs, errors.Wrapf(err, "error parsing %s", hdr.Name))
-						return true, nil
-					}
-					o.Manifests = append(o.Manifests, ms...)
-				}
-				return true, nil
-			}
-		}
-		if err := opts.Run(); err != nil {
-			return err
-		}
-		if !found {
-			return fmt.Errorf("image did not contain %s", o.File)
-		}
-
-		// Only output manifest errors if manifests were being extracted.
-		// Do not return an error so current operation, e.g. mirroring, continues.
-		if o.ExtractManifests && len(manifestErrs) > 0 {
-			fmt.Fprintf(o.ErrOut, "Errors: %s\n", errorList(manifestErrs))
-		}
-		return nil
+	tarEntryCallbacks := []extract.TarEntryFunc{}
 
-	case o.CredentialsRequests:
+	var manifestErrs []error
+	if o.ExtractManifests {
 		expectedProviderSpecKind := ""
-		if len(o.Cloud) > 0 {
+		if o.Cloud != "" {
 			expectedProviderSpecKind = credRequestCloudProviderSpecKindMapping[o.Cloud]
 		}
-		opts.TarEntryCallback = func(hdr *tar.Header, _ extract.LayerInfo, r io.Reader) (bool, error) {
+
+		tarEntryCallbacks = append(tarEntryCallbacks, func(hdr *tar.Header, _ extract.LayerInfo, r io.Reader) (bool, error) {
 			if ext := path.Ext(hdr.Name); len(ext) == 0 || !(ext == ".yaml" || ext == ".yml" || ext == ".json") {
 				return true, nil
 			}
 			klog.V(4).Infof("Found manifest %s", hdr.Name)
-			raw, err := ioutil.ReadAll(r)
+			raw, err := io.ReadAll(r)
```

Comment:
this is not new code but I wonder if we could simplify and `manifest.ParseManifest(r)` instead of reading it to memory and then making a reader from it again

Reply:
Done in 6e980ae0d -> b2b34df0a.